src/compute/src/render/continual_task.rs

```diff
@@ -143,50 +249,28 @@ impl<G: Scope<Timestamp = Timestamp>> ContinualTaskCtx<G> {
     }
 
     pub fn is_ct_dataflow(&self) -> bool {
-        !self.ct_inputs.is_empty()
+        // Inputs are non-empty iff outputs are non-empty.
+        assert_eq!(self.ct_inputs.is_empty(), self.ct_outputs.is_empty());
+        !self.ct_outputs.is_empty()
     }
 
-    pub fn get_ct_inserts_transformer<S: Scope<Timestamp = Timestamp>>(
+    pub fn get_ct_source_transformer(
         &self,
         source_id: GlobalId,
-    ) -> Option<
-        impl FnOnce(
-            Collection<S, Row, Diff>,
-            Collection<S, DataflowError, Diff>,
-        ) -> (
-            Collection<S, Row, Diff>,
-            Collection<S, DataflowError, Diff>,
-            Collection<S, (), Diff>,
-        ),
-    > {
-        if !self.ct_inputs.contains(&source_id) {
+    ) -> Option<ContinualTaskSourceTransformer> {
+        if !self.is_ct_dataflow() {
             return None;
         }
-
-        // Make a collection s.t, for each time T in the input, the output
-        // contains the inserts at T.
-        let inserts_source_fn =
-            move |oks: Collection<S, Row, Diff>, errs: Collection<S, DataflowError, Diff>| {
-                let name = source_id.to_string();
-                // Keep only the inserts.
-                //
-                // At some point this will become a user option to instead keep
-                // only deletes.
-                let oks = oks.inner.filter(|(_, _, diff)| *diff > 0);
-                // Grab the original times for use in the sink operator.
-                let times = oks.map(|(_row, ts, diff)| ((), ts, diff));
-                // SUBTLE: See the big module rustdoc for what's going on here.
-                let oks = step_backward(&name, oks.as_collection());
-                let errs = step_backward(&name, errs);
-                // Then retract everything at the next timestamp.
-                let oks = oks.inner.flat_map(|(row, ts, diff)| {
-                    let mut negation = diff.clone();
-                    differential_dataflow::Diff::negate(&mut negation);
-                    [(row.clone(), ts.step_forward(), negation), (row, ts, diff)]
-                });
-                (oks.as_collection(), errs, times.as_collection())
-            };
-        Some(inserts_source_fn)
+        let transformer = match (
+            self.ct_inputs.contains(&source_id),
+            self.ct_outputs.contains(&source_id),
+        ) {
+            (false, false) => ContinualTaskSourceTransformer::NormalReference,
+            (false, true) => ContinualTaskSourceTransformer::SelfReference { source_id },
+            (true, false) => ContinualTaskSourceTransformer::InsertsInput,
+            (true, true) => unreachable!("ct output is not allowed to be an input"),
```

Comment:
Feels like this should be a `panic`? At least I think of `unreachable` as marking branches that can be locally reasoned about as unreachable. I don't think that's the case here since nothing prevents us from constructing a dataflow description that specifies the same ID for the input and the output.